api/src/main/java/com/epam/pipeline/manager/datastorage/providers/aws/s3/S3Helper.java

```diff
@@ -300,6 +302,54 @@ public void restoreFileVersion(String bucket, String path, String version) {
         moveS3Object(client, bucket, new MoveObjectRequest(path, version, path));
     }
 
+    public void restoreFolder(final String bucket, final String path, final RestoreFolderVO restoreFolderVO) {
+        final AmazonS3 client = getDefaultS3Client();
+        final String requestPath = Optional.ofNullable(path).orElse("");
+        if (!StringUtils.isNullOrEmpty(requestPath)) {
+            Assert.isTrue(checkItemType(client, bucket, requestPath, true) == DataStorageItemType.Folder,
+                    messageHelper.getMessage(MessageConstants.ERROR_FOLDER_INVALID_PATH, requestPath));
+        }
+        try (S3ObjectDeleter deleter = new S3ObjectDeleter(client, bucket)) {
+            restoreFolderFiles(client, bucket, requestPath, restoreFolderVO, deleter);
+        } catch (SdkClientException e) {
+            throw new DataStorageException(e.getMessage(), e.getCause());
+        }
+    }
+
+    private void restoreFolderFiles(final AmazonS3 client, final String bucket, final String path,
+                                    final RestoreFolderVO restoreFolderVO, final S3ObjectDeleter deleter) {
+        getDeletedFiles(client, bucket, ProviderUtils.withTrailingDelimiter(path), restoreFolderVO, deleter)
+                .forEach(item -> deleter.deleteKey(item.getPath(), ((DataStorageFile) item).getVersion()));
```

Comment:
With current approach we collect all delete markers (not files BTW) at first and then try to delete all of them as the last step. This may lead to OOM for large folders. I'd suggest to modify method `getDeletedFiles` to `cleanDeleteMarkers` (or similar) and perform deletion in it.